File: examples/py/tokitai_client.py
"""Tokitai Python SDK client.

A minimal async client for the Tokitai MCP-compatible HTTP server.

The server exposes three endpoints (see docs/CROSS_LANGUAGE.md):
    GET  /tools       -> list of McpTool definitions
    POST /call        -> invoke a tool, returns {success, result, error}
    GET  /sse/<name>  -> Server-Sent Events stream of tool output (optional)

Only the standard MCP-shaped subset is required. The SSE endpoint is
feature-gated: if the server does not expose it, ``stream_tool`` raises
a clear ``NotImplementedError`` rather than silently hanging.
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, AsyncIterator, Mapping

import httpx


@dataclass
class McpTool:
    """Subset of the McpTool JSON shape that callers care about."""

    name: str
    description: str
    input_schema: dict[str, Any]

    @classmethod
    def from_json(cls, data: Mapping[str, Any]) -> "McpTool":
        return cls(
            name=data["name"],
            description=data.get("description", ""),
            input_schema=data.get("input_schema", {}),
        )


class TokitaiClient:
    """Async client for a Tokitai MCP server.

    Parameters
    ----------
    base_url:
        Origin of the server, e.g. ``"http://127.0.0.1:8080"``. Trailing
        slashes are tolerated.
    timeout:
        Default request timeout in seconds for unary calls. Streaming
        calls ignore this and use a long-lived connection.
    """

    def __init__(self, base_url: str, *, timeout: float = 30.0) -> None:
        self.base_url = base_url.rstrip("/")
        self._timeout = timeout
        # One shared client keeps connection pooling cheap.
        self._http = httpx.AsyncClient(base_url=self.base_url, timeout=timeout)

    async def __aenter__(self) -> "TokitaiClient":
        return self

    async def __aexit__(self, *exc: Any) -> None:
        await self.aclose()

    async def aclose(self) -> None:
        await self._http.aclose()

    # ------------------------------------------------------------------
    # Discovery
    # ------------------------------------------------------------------
    async def list_tools(self) -> list[McpTool]:
        """Return every tool the server has registered."""
        resp = await self._http.get("/tools")
        resp.raise_for_status()
        return [McpTool.from_json(item) for item in resp.json()]

    # ------------------------------------------------------------------
    # Unary invocation
    # ------------------------------------------------------------------
    async def call_tool(self, name: str, arguments: Mapping[str, Any] | None = None) -> dict[str, Any]:
        """Invoke ``name`` with ``arguments`` and return the raw response dict.

        The response always has the keys ``success`` (bool) and, on success,
        ``result`` (any JSON value) or, on failure, ``error`` (string). This
        method does not raise on tool-level errors — the caller is expected
        to inspect ``success``. HTTP-level errors (4xx/5xx) still raise.
        """
        payload = {"name": name, "arguments": dict(arguments or {})}
        resp = await self._http.post("/call", json=payload)
        resp.raise_for_status()
        return resp.json()

    # ------------------------------------------------------------------
    # Streaming (SSE)
    # ------------------------------------------------------------------
    async def stream_tool(
        self,
        name: str,
        arguments: Mapping[str, Any] | None = None,
    ) -> AsyncIterator[dict[str, Any]]:
        """Yield incremental results from the SSE endpoint.

        The server may not be built with SSE support. We probe ``/sse/<name>``
        with a HEAD request first and raise ``NotImplementedError`` if the
        endpoint is absent (404) so callers fail fast instead of waiting on
        a connection that will never deliver data.
        """
        # Probe for SSE support.
        probe = await self._http.head(f"/sse/{name}")
        if probe.status_code == httpx.codes.NOT_FOUND:
            raise NotImplementedError(
                f"Server {self.base_url!r} does not expose SSE streaming. "
                "Rebuild tokitai-mcp-server with the 'sse' feature or fall "
                "back to call_tool()."
            )

        url = f"/sse/{name}"
        # SSE uses long-lived chunked responses; disable the read timeout.
        async with self._http.stream(
            "POST", url, json={"name": name, "arguments": dict(arguments or {})}
        ) as resp:
            resp.raise_for_status()
            async for line in resp.aiter_lines():
                if not line or not line.startswith("data:"):
                    continue
                payload = line[len("data:"):].strip()
                if not payload:
                    continue
                try:
                    yield json.loads(payload)
                except json.JSONDecodeError:
                    yield {"raw": payload}


# ---------------------------------------------------------------------------
# Manual smoke test
# ---------------------------------------------------------------------------
async def _demo() -> None:  # pragma: no cover - illustrative
    """Print the call sequence for the canonical ``add(1, 2)`` example.

    This function is intentionally a no-network demo: it just shows the
    sequence a caller would issue. Run the Rust example server in another
    terminal first:

        cargo run -p tokitai-mcp-server --example mcp_builder_demo
    """
    print("Tokitai Python client demo")
    print("  1. GET  http://127.0.0.1:8080/tools")
    print("  2. POST http://127.0.0.1:8080/call  {name: 'add', arguments: {a: 1, b: 2}}")
    print("  -> expect {success: True, result: 3}")


if __name__ == "__main__":
    import asyncio
    asyncio.run(_demo())

File: examples/js/tokitai-client.ts
/**
 * Tokitai TypeScript client.
 *
 * A typed, zero-runtime-dependency client for the Tokitai MCP-compatible
 * HTTP server. Uses only Web APIs (`fetch`) so it runs in Node 18+,
 * browsers, Deno, Bun, and edge runtimes.
 *
 * Server endpoints (see docs/CROSS_LANGUAGE.md):
 *   GET  /tools       -> McpTool[]
 *   POST /call        -> ToolCallResponse
 *   POST /sse/<name>  -> Server-Sent Events (optional)
 */

// ---------------------------------------------------------------------------
// Types — strict shapes matching the Rust server's JSON.
// ---------------------------------------------------------------------------

export interface McpTool {
  readonly name: string;
  readonly description: string;
  readonly input_schema: Record<string, unknown>;
}

export interface ToolCallRequest {
  readonly name: string;
  readonly arguments: Record<string, unknown>;
}

export interface ToolCallResponse {
  readonly success: boolean;
  readonly result?: unknown;
  readonly error?: string;
}

export interface StreamEvent {
  readonly data: unknown;
  readonly event?: string;
  readonly id?: string;
}

// ---------------------------------------------------------------------------
// Client
// ---------------------------------------------------------------------------

export interface TokitaiClientOptions {
  /** Request timeout in milliseconds for unary calls. Default: 30 000. */
  readonly timeoutMs?: number;
  /** Custom fetch implementation (handy for tests / Cloudflare Workers). */
  readonly fetchImpl?: typeof fetch;
}

export class TokitaiError extends Error {
  constructor(message: string, readonly cause?: unknown) {
    super(message);
    this.name = "TokitaiError";
  }
}

export class SseNotSupportedError extends TokitaiError {
  constructor(baseUrl: string) {
    super(
      `Server ${baseUrl} does not expose SSE streaming. ` +
        "Rebuild tokitai-mcp-server with the 'sse' feature or fall back to callTool().",
    );
    this.name = "SseNotSupportedError";
  }
}

export class TokitaiClient {
  readonly baseUrl: string;
  private readonly timeoutMs: number;
  private readonly fetchImpl: typeof fetch;

  constructor(baseUrl: string, opts: TokitaiClientOptions = {}) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.timeoutMs = opts.timeoutMs ?? 30_000;
    this.fetchImpl = opts.fetchImpl ?? fetch;
  }

  // -----------------------------------------------------------------
  // Discovery
  // -----------------------------------------------------------------
  async listTools(): Promise<McpTool[]> {
    const resp = await this.request("GET", "/tools");
    if (!resp.ok) {
      throw new TokitaiError(`listTools failed: ${resp.status} ${resp.statusText}`);
    }
    return (await resp.json()) as McpTool[];
  }

  // -----------------------------------------------------------------
  // Unary invocation
  // -----------------------------------------------------------------
  async callTool(name: string, args: Record<string, unknown> = {}): Promise<ToolCallResponse> {
    const body: ToolCallRequest = { name, arguments: args };
    const resp = await this.request("POST", "/call", body);
    if (!resp.ok) {
      throw new TokitaiError(`callTool failed: ${resp.status} ${resp.statusText}`);
    }
    return (await resp.json()) as ToolCallResponse;
  }

  // -----------------------------------------------------------------
  // Streaming (SSE)
  // -----------------------------------------------------------------
  /**
   * Subscribe to the SSE stream for `name`. The server's response must
   * use `text/event-stream`; otherwise we throw `SseNotSupportedError`
   * so callers fail fast instead of buffering forever.
   *
   * Implemented with `fetch` + a local SSE parser, so it works in any
   * environment that ships a global `fetch` (Node 18+, browsers, Deno,
   * Bun) — no `EventSource` polyfill is required.
   */
  async *streamTool(
    name: string,
    args: Record<string, unknown> = {},
  ): AsyncGenerator<StreamEvent> {
    const resp = await this.request("POST", `/sse/${encodeURIComponent(name)}`, {
      name,
      arguments: args,
    });
    if (resp.status === 404) {
      throw new SseNotSupportedError(this.baseUrl);
    }
    if (!resp.ok || !resp.body) {
      throw new TokitaiError(`streamTool failed: ${resp.status} ${resp.statusText}`);
    }

    yield* parseSseStream(resp.body);
  }

  // -----------------------------------------------------------------
  // Internals
  // -----------------------------------------------------------------
  private request(
    method: string,
    path: string,
    body?: unknown,
  ): Promise<Response> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    const init: RequestInit = {
      method,
      headers: body !== undefined ? { "content-type": "application/json" } : undefined,
      body: body !== undefined ? JSON.stringify(body) : undefined,
      signal: controller.signal,
    };
    return this.fetchImpl(`${this.baseUrl}${path}`, init).finally(() =>
      clearTimeout(timer),
    );
  }
}

// ---------------------------------------------------------------------------
// SSE parsing
// ---------------------------------------------------------------------------

/**
 * Minimal SSE parser operating on a WHATWG ReadableStream. Yields one
 * `StreamEvent` per `data:` block. Multi-line `data:` fields are joined
 * with newlines per the SSE spec.
 */
export async function* parseSseStream(
  body: ReadableStream<Uint8Array>,
): AsyncGenerator<StreamEvent> {
  const decoder = new TextDecoder("utf-8");
  const reader = body.getReader();
  let buffer = "";
  let dataLines: string[] = [];
  let eventName: string | undefined;
  let id: string | undefined;

  const dispatch = (): StreamEvent | undefined => {
    if (dataLines.length === 0 && !eventName && !id) return undefined;
    const raw = dataLines.join("\n");
    let parsed: unknown = raw;
    if (raw.length > 0) {
      try {
        parsed = JSON.parse(raw);
      } catch {
        // Keep as raw string.
      }
    }
    const evt: StreamEvent = { data: parsed, event: eventName, id };
    dataLines = [];
    eventName = undefined;
    id = undefined;
    return evt;
  };

  try {
    for (;;) {
      const { value, done } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      let nl: number;
      while ((nl = buffer.indexOf("\n")) >= 0) {
        const line = buffer.slice(0, nl).replace(/\r$/, "");
        buffer = buffer.slice(nl + 1);
        if (line === "") {
          const evt = dispatch();
          if (evt) yield evt;
          continue;
        }
        if (line.startsWith(":")) continue; // comment
        const colon = line.indexOf(":");
        const field = colon === -1 ? line : line.slice(0, colon);
        let val = colon === -1 ? "" : line.slice(colon + 1);
        if (val.startsWith(" ")) val = val.slice(1);
        switch (field) {
          case "data":
            dataLines.push(val);
            break;
          case "event":
            eventName = val;
            break;
          case "id":
            id = val;
            break;
        }
      }
    }
    const tail = dispatch();
    if (tail) yield tail;
  } finally {
    reader.releaseLock();
  }
}
